build-system/babel-plugins/babel-plugin-transform-function-declarations/index.js

```diff
@@ -14,119 +14,198 @@
+
+  const EXPRESSION_BAIL_OUT_CONDITIONS = [
+    isNotFunction,
+    isGenerator,
+    isNotSingleReturnStatement,
+    containsArgumentsUsage,
+    containsThisExpression,
+    (path) => {
+      const callExpression = path.findParent((p) => p.isCallExpression());
+      if (callExpression) {
+        const {name} = (callExpression.node && callExpression.node.callee) || {
+          name: null,
+        };
+        return EXHEMPTED_EXPRESSION_NAME_REGEXS.every((regexp) =>
+          regexp.test(name)
+        );
+      }
+
+      const declarator = path.findParent((p) => p.isVariableDeclarator());
+      if (declarator) {
+        return referencesAreOnlyCallExpressions(path, declarator.node.id.name);
+      }
+
+      return true;
+    },
+  ];
+
   return {
     name: 'safe-arrows',
     visitor: {
-      FunctionDeclaration(path) {
-        for (const [id, method] of Object.entries(BAIL_OUT_CONDITIONS)) {
-          if (method(path)) {
-            if (DEBUG) {
-              const pathId = path.get('id');
-              const message = pathId
-                ? `Bail on function ${pathId.node.name}: Reason ${id}.`
-                : `Bail on item: Reason ${id}`;
-              console /*OK*/
-                .log(message);
+      FunctionExpression(path) {
+        const insideCallExpression = path.findParent((p) =>
+          p.isCallExpression()
+        );
+        if (insideCallExpression) {
+          for (const method of EXPRESSION_BAIL_OUT_CONDITIONS) {
+            if (method(path)) {
+              return;
             }
+          }
+          path.replaceWith(createArrowFunctionExpression(path.node));
+        } else {
+          const declarator = path.findParent((p) => p.isVariableDeclarator());
+          if (!declarator) {
             return;
           }
-        }
 
-        if (DEBUG) {
-          const {name} = path.node.id;
-          console /*OK*/
-            .log(`Success for function ${name}.`);
+          for (const method of EXPRESSION_BAIL_OUT_CONDITIONS) {
+            if (method(path)) {
+              return;
+            }
+          }
+          const name =
+            (declarator.node &&
+              declarator.node.id &&
+              declarator.node.id.name) ||
+            null;
+          declarator.replaceWith(createVariableDeclarator(path, name));
         }
-
+      },
+      FunctionDeclaration(path) {
         const usableBlock = path.findParent(
           (path) => path.isBlockStatement() || path.isProgram()
         );
-        if (usableBlock) {
-          usableBlock.unshiftContainer('body', createVariableDeclaration(path));
-          path.remove();
+
+        if (!usableBlock) {
+          return;
         }
+
+        for (const method of DECLARATION_BAIL_OUT_CONDITIONS) {
+          if (method(path)) {
+            return;
+          }
+        }
+
+        // Making an existing FunctionDeclaration a let VariableDeclaration is safe.
```

Comment:
do we insert the let declaration to the top of the current scope? (in case usage is before the original declaration point)